### DISCUSSION TITLE
"_p": "F" in logs

### DISCUSSION THREAD

**Discussion by prabhatsharma** [fluent/fluent-bit#7707] [Q&A] [answered]
I see that all the logs sent by fluentbit from my k8s have 
```
"_p": "F",
```

What does this mean?

**Comment by agup006** [comment, marked as answer]
@prabhatsharma 

This is from CRI and Docker logging formats indicating partial and full logs. These are mostly used by Fluent Bit when performing multiline logging via the cri, docker parsers.

Here is a quick excerpt of the format docs https://github.com/kubernetes/design-proposals-archive/blob/main/node/kubelet-cri-logging.md

```
Log format

The runtime should decorate each log entry with a RFC 3339Nano timestamp prefix, the stream type (i.e., "stdout" or "stderr"), the tags of the log entry, the log content that ends with a newline.

The tags fields can support multiple tags, delimited by :. Currently, only one tag is defined in CRI to support multi-line log entries: partial or full. Partial (P) is used when a log entry is split into multiple lines by the runtime, and the entry has not ended yet. Full (F) indicates that the log entry is completed -- it is either a single-line entry, or this is the last line of the multiple-line entry.

For example,

2016-10-06T00:17:09.669794202Z stdout F The content of the log entry 1
2016-10-06T00:17:09.669794202Z stdout P First line of log entry 2
2016-10-06T00:17:09.669794202Z stdout P Second line of the log entry 2
2016-10-06T00:17:10.113242941Z stderr F Last line of the log entry 2
With the knowledge, kubelet can parse the logs and serve them for kubectl logs requests. This meets requirement (3). Note that the format is defined deliberately simple to provide only information necessary to serve the requests. We do not intend for kubelet to host various logging plugins. It is also worth mentioning again that the scope of this proposal is restricted to stdout/stderr streams of the container, and we impose no restriction to the logging format of arbitrary container logs.
```

**Reply by AndrewSav** [reply]
those parsers aka multiline parsers, are supposed to merge the lines split by the engine together. when they do, a sequence of "P, .., P, F" is to become a single line. So after the merge it should be impossible to distinguish "P" and "F" because they all will be merged together. And yet, we are seeing `"_p": "F",` and `"_p": "P"`. What is their meaning after the merge? @agup006

**Reply by patrick-stephens** [reply]
Actual examples will help along with full config. It's really hard to answer generically because it may be misconfiguration, invalid input or something else.

**Reply by AndrewSav** [reply]
@patrick-stephens let's assume there was not misconfiguration or invalid input. Are you saying that in this circumstances one should not be seeing 

```
"_p": "F",
```

coming through? 

Just use `multiline.parser cri` with `tail` input for kubernetes pods and you will see these in the output, there rest of the config does not seem to make any difference. And this above is the actual example, the rest of the data is custom and do not appear to make any difference - they can be anything, we will still see this field. The question is, **what is the meaning and intention of the field** after being processed by FluentBit with regards to multiline logging?

Here is the the config:

```
[SERVICE]
    Daemon Off
    Flush 1
    Log_Level info
    Parsers_File /fluent-bit/etc/parsers.conf
    Parsers_File /fluent-bit/etc/conf/custom_parsers.conf
    HTTP_Server On
    HTTP_Listen 0.0.0.0
    HTTP_Port 2020
    Health_Check On
[INPUT]
    Name tail
    Tag kube.*
    Path /var/log/containers/*.log
    Exclude_Path *fluentbit*
    Mem_Buf_Limit 50MB
    Rotate_Wait 30
    multiline.parser cri
[FILTER]
    Name kubernetes
    Match kube.*
[FILTER]
    Name record_modifier
    Match kube.*
    Record cluster staging2
[FILTER]
    Name          rewrite_tag
    Match         kube.*
    Rule          $log .* crowdstrike_fork true
[FILTER]
    name   grep
    match  kube.*
    regex  $kubernetes['labels']['btlog'] ^enabled$
[FILTER]
    Name record_modifier
    Match kube.*
    Record btlog true
[FILTER]
    name   grep
    match  crowdstrike_fork
    regex  $kubernetes['labels']['btsiem'] ^enabled$
[OUTPUT]
    Name                    gelf
    Match                   kube.*
    Host                    ${FLUENT_GELF_HOST}
    Port                    ${FLUENT_GELF_PORT}
    Mode                    ${FLUENT_GELF_PROTOCOL}
    Gelf_Short_Message_Key  log
[OUTPUT]
    name syslog
    match crowdstrike_fork
    host ${FLUENT_SYSLOG_HOST}
    port ${FLUENT_SYSLOG_PORT}
    mode tcp
    syslog_maxsize 8000
    syslog_message_key log
    net.dns.mode            TCP
    net.dns.prefer_ipv4     true
```


And here is the data:
![2025-07-10_09-38-09a](https://github.com/user-attachments/assets/cdbabc95-6ea9-4c35-a5ce-ced6285f5c33)

This is screenshot from Seq where data is injested from FluentBit with GELF plugin, but I can also see the field in question when I use the `stdout` filter.
